File: chapter3/java/src/main/java/v1/Transcriber.java
package v1;

import common.DnaSequence;
import common.RnaBase;
import common.RnaSequence;

import java.util.ArrayList;

public class Transcriber {

    RnaSequence transcribe(DnaSequence sequence) {
        var out = new ArrayList<RnaBase>();
        for (var dna : sequence) {
            RnaBase rna = null;
            switch (dna) {
                case C: rna = RnaBase.G;
                    break;
                case G: rna = RnaBase.C;
                    break;
                case A: rna = RnaBase.U;
                    break;
                case T: rna = RnaBase.A;
                    break;
            }
            out.add(rna);
        }
        return new RnaSequence(out);
    }

}

File: chapter3/csharp/V1/Transcriber.cs
using System.Linq;

namespace V1
{
    public class Transcriber
    {
        private RnaBase PairOf(DnaBase dnaBase)
        {
            switch (dnaBase)
            {
                case DnaBase.A: return RnaBase.U;
                case DnaBase.C: return RnaBase.G;
                case DnaBase.G: return RnaBase.C;
                case DnaBase.T: return RnaBase.A;
                default: return RnaBase.A;
            }
        }

        public RnaSequence Transcribe(DnaSequence dnaSequence)
        {
            return new RnaSequence(
                dnaSequence
                    .Select(dna => PairOf(dna))
                    .ToList());
        }
    }
}